File: backend/api/routes/traces.py
"""Trace observability API routes."""

from fastapi import APIRouter

from backend.api.errors import NotFoundError, ServiceUnavailableError

router = APIRouter()


def _get_observability():
    from backend.pipeline.observability.manager import get_active_manager
    mgr = get_active_manager()
    if not mgr:
        raise ServiceUnavailableError("Observability not enabled", hint="Enable observability in platform configuration")
    return mgr


@router.get(
    "/summary",
    summary="Trace summary",
    description="Summary of all in-memory traces.",
)
async def trace_summary():
    """Get summary of all in-memory traces.

    Returns the documented contract plus the additive fields the trace
    viewer consumes:

        {
          "total_traces": 42,        # documented
          "active_traces": 3,        # documented; trace with span activity
                                     # within the last 5 minutes
          "error_rate": 0.05,        # documented; error spans / total spans
          "recent_traces": [         # additive; newest first, bounded (50)
            {
              "trace_id": "abc123...",
              "span_count": 12,
              "started_at": 1726100000.0,
              "last_activity": 1726100120.0,
              "duration_ms": 120000.0,
              "error_count": 0,
              "models": ["glm-5.2"]   # providers seen via cost linkage
            }
          ],
          "span_count": 500, "trace_count": 42, "by_kind": {...},
          "by_status": {...}, "total_duration_ms": ..., "avg_duration_ms": ...
        }

    The legacy aggregate keys (span_count, trace_count, by_kind, by_status,
    total_duration_ms, avg_duration_ms) are preserved for existing consumers.
    """
    return _get_observability().get_trace_summary()


@router.get(
    "/trace/{trace_id}",
    summary="Get trace spans",
    description="Get all spans for a specific trace by its ID.",
)
async def get_trace(trace_id: str):
    """Get all spans for a specific trace.

    Args:
        trace_id: The unique trace identifier.

    Returns:
        {"trace_id": "...", "spans": [...]}

    Example response:
        {"trace_id": "abc-123", "spans": [{"name": "generation", "duration_ms": 1500}]}
    """
    spans = _get_observability().get_traces(trace_id)
    if not spans:
        raise NotFoundError(f"No spans found for trace {trace_id}")
    return {"trace_id": trace_id, "spans": spans}


@router.get(
    "/metrics",
    summary="Trace metrics",
    description="Current metrics snapshot including latency percentiles and error rates.",
)
async def trace_metrics():
    """Get current metrics snapshot.

    Returns the documented flat aggregate plus the per-kind breakdown:

        {
          "p50_ms": 120, "p95_ms": 800, "p99_ms": 3500, "avg_ms": 210.5,
          "error_rate": 0.02, "call_count": 128,
          "by_kind": {"stage": {"count": ..., "latency_ms": {...}}, ...}
        }
    """
    return _get_observability().get_metrics()

File: frontend/src/api/traces.ts
/**
 * Traces API Client — BATCH-21/TASK-01
 *
 * Typed functions for trace observability endpoints.
 * Endpoint shapes from backend/api/routes/traces.py:
 *   GET /traces/summary     → {total_traces, active_traces, error_rate}
 *   GET /traces/trace/{id}  → {trace_id, spans: [{name, duration_ms, ...}]}
 *   GET /traces/metrics     → {p50_ms, p99_ms, error_rate}
 */

import { callContract } from "./contracts/common";
import {
  getTraceContract,
  getTraceMetricsContract,
  getTraceSummaryContract,
} from "./contracts/traces";

// ── Types ────────────────────────────────────────────────────────

/** One recent trace from GET /traces/summary — real id, real stats. */
export interface RecentTrace {
  trace_id: string;
  span_count: number;
  /** Epoch seconds of the trace's earliest span start. */
  started_at?: number;
  /** Epoch seconds of the trace's most recent span end. */
  last_activity?: number;
  duration_ms: number;
  error_count?: number;
  /** Provider models observed on the trace's spans via cost linkage. */
  models?: string[];
}

export interface TraceSummary {
  total_traces: number;
  active_traces: number;
  error_rate: number;
  recent_traces: RecentTrace[];
}

export interface TraceSpan {
  name: string;
  duration_ms: number;
  [key: string]: unknown;
}

export interface TraceDetail {
  trace_id: string;
  spans: TraceSpan[];
}

export interface TraceMetrics {
  p50_ms: number;
  p99_ms: number;
  error_rate: number;
  p95_ms?: number;
  avg_ms?: number;
  call_count?: number;
  by_kind?: Record<string, unknown>;
}

// ── API Functions ────────────────────────────────────────────────

export function getTraceSummary(): Promise<TraceSummary> {
  // F1.7a: migrated from apiFetchUnchecked to callContract with runtime decoder
  return callContract(getTraceSummaryContract);
}

export function getTrace(traceId: string): Promise<TraceDetail> {
  // F1.7a: migrated from apiFetchUnchecked to callContract with runtime decoder
  return callContract(getTraceContract, { params: { traceId } });
}

export function getTraceMetrics(): Promise<TraceMetrics> {
  // F1.7a: migrated from apiFetchUnchecked to callContract with runtime decoder
  return callContract(getTraceMetricsContract);
}
